File: api/src/sim/world.ts
/**
 * The world model, and how a battle's opening state is built from roster data.
 *
 * The roster is a **multiset** (design doc §4.1): "3× Ember Adept, 4× Cinder
 * Hound" is the normal shape, and nothing in the sim assumes an army size. Round
 * 1 fields three mages and round 4 fields most of the roster, so any code here
 * that hard-coded a count would be wrong by round 2.
 *
 * A troop is one or more mages plus the summons they support (§4.2). Support is
 * mandatory and local — a mage in another troop is too far away to help — so a
 * summon whose school no mage in *its own* troop supports is rejected at build
 * time rather than quietly standing on the field.
 *
 * Placement here is the flat "everyone into the strip" version. Formations
 * derived from orders are slice B (JQ-287); this is what it replaces.
 */
import type { MapConfig } from './map.js';
import { validateMapConfig } from './map.js';
import type { Rng, RngState } from './rng.js';
import type { School } from './schools.js';
import type { Side, TroopId, UnitId, UnitRef, Vec2 } from './types.js';
import { SIDES } from './types.js';
import { buildUnitTypeCatalog, type UnitKind, type UnitType, type UnitTypeCatalog } from './units.js';

/** One line of a roster: a card and how many copies of it. Omitting `count` means one. */
export interface RosterEntry {
  typeId: string;
  count?: number;
}

export interface TroopSetup {
  id?: string;
  mages: RosterEntry[];
  summons: RosterEntry[];
}

export interface ArmySetup {
  side: Side;
  troops: TroopSetup[];
}

/** Everything a battle needs to exist, as data. */
export interface BattleSetup {
  unitTypes: UnitType[];
  armies: ArmySetup[];
}

export interface Unit {
  readonly id: UnitId;
  readonly typeId: string;
  readonly kind: UnitKind;
  readonly schools: readonly School[];
  readonly side: Side;
  readonly troopId: TroopId;
  readonly maxHp: number;
  readonly damage: number;
  readonly range: number;
  /** Map units per second; the loop converts to per-tick. */
  readonly speed: number;
  readonly attackCooldownSeconds: number;
  hp: number;
  position: Vec2;
  /** Ticks still to wait before this unit can attack again. Counted in ticks, not seconds — see `toTicks`. */
  cooldownRemaining: number;
}

export interface Troop {
  readonly id: TroopId;
  readonly side: Side;
  /** Living mages. A troop whose last mage dies dissolves — slice D (§4.6). */
  mageIds: UnitId[];
  summonIds: UnitId[];
}

export interface BaseState {
  readonly maxHp: number;
  readonly position: Vec2;
  hp: number;
}

export interface World {
  tick: number;
  /** Serialisable: a snapshot fully determines every draw that follows it. */
  rngState: RngState;
  units: Unit[];
  troops: Troop[];
  bases: Record<Side, BaseState>;
  zoneScore: Record<Side, number>;
}

/** How a unit appears in the event stream. */
export function unitRef(unit: Unit): UnitRef {
  return { unitId: unit.id, troopId: unit.troopId, side: unit.side, typeId: unit.typeId };
}

/** Spacing between deployed units, in map units. Sprites are 18–28 px (JQ-243). */
const DEPLOY_SPACING = 24;
/** A unit of jitter, so placement reads as an army rather than a spreadsheet. */
const DEPLOY_JITTER = 1;

function expand(entries: readonly RosterEntry[], catalog: UnitTypeCatalog, kind: UnitKind): UnitType[] {
  const expanded: UnitType[] = [];

  for (const entry of entries) {
    const type = catalog[entry.typeId];
    if (!type) {
      throw new Error(`roster names ${entry.typeId}, which is not in the unit type catalog`);
    }
    if (type.kind !== kind) {
      throw new Error(`${entry.typeId} is a ${type.kind}, but the roster lists it as a ${kind}`);
    }
    const count = entry.count ?? 1;
    if (!Number.isInteger(count) || count < 1) {
      throw new Error(`roster entry ${entry.typeId} has a count of ${count}`);
    }
    for (let copy = 0; copy < count; copy += 1) {
      expanded.push(type);
    }
  }

  return expanded;
}

function assertOneArmyPerSide(armies: readonly ArmySetup[]): void {
  const seen = new Set<Side>();
  for (const army of armies) {
    if (seen.has(army.side)) {
      throw new Error(`two armies were given for ${army.side}`);
    }
    seen.add(army.side);
  }
  for (const side of SIDES) {
    if (!seen.has(side)) {
      throw new Error(`no army was given for ${side}`);
    }
  }
}

function assertSupported(mages: readonly UnitType[], summons: readonly UnitType[], troopId: TroopId): void {
  if (mages.length === 0) {
    throw new Error(`troop ${troopId} has no mage, so it could hold nothing on the field`);
  }

  const supported = new Set<School>(mages.flatMap((mage) => mage.schools));
  for (const summon of summons) {
    if (!summon.schools.some((school) => supported.has(school))) {
      throw new Error(
        `troop ${troopId} has no mage able to support ${summon.id} ` +
          `(${summon.schools.join('/')}); support is mandatory and local`,
      );
    }
  }
}

function instantiate(
  type: UnitType,
  id: UnitId,
  side: Side,
  troopId: TroopId,
  position: Vec2,
): Unit {
  return {
    id,
    typeId: type.id,
    kind: type.kind,
    schools: type.schools,
    side,
    troopId,
    maxHp: type.maxHp,
    damage: type.damage,
    range: type.range,
    speed: type.speed,
    attackCooldownSeconds: type.attackCooldownSeconds,
    hp: type.maxHp,
    position,
    cooldownRemaining: 0,
  };
}

function clamp(value: number, low: number, high: number): number {
  return Math.min(high, Math.max(low, value));
}

/**
 * Lays a side's units out in its deployment strip, filling columns before adding
 * a row — a row costs 21 px of the scarce portrait axis, a column 22 px of width
 * the map was not using (JQ-243).
 */
function deploymentSpot(map: MapConfig, side: Side, index: number, rng: Rng): Vec2 {
  const strip = map.deployment[side];
  const width = strip.extent.to - strip.extent.from;
  const columns = Math.max(1, Math.floor(width / DEPLOY_SPACING));
  const column = index % columns;
  const row = Math.floor(index / columns);

  const x = strip.extent.from + DEPLOY_SPACING / 2 + column * DEPLOY_SPACING;
  const jitter = (rng.nextFloat() * 2 - 1) * DEPLOY_JITTER;

  // Front rank nearest the zones; further ranks fall back toward the base.
  const depth = DEPLOY_SPACING / 2 + row * DEPLOY_SPACING;
  const y = side === 'north' ? strip.lane.to - depth : strip.lane.from + depth;

  return {
    x: clamp(x + jitter, strip.extent.from, strip.extent.to),
    y: clamp(y, strip.lane.from, strip.lane.to),
  };
}

/** Builds the opening state of a battle. Pure: the same inputs give the same world. */
export function createWorld(map: MapConfig, battleState: BattleSetup, rng: Rng): World {
  validateMapConfig(map);
  assertOneArmyPerSide(battleState.armies);

  const catalog = buildUnitTypeCatalog(battleState.unitTypes);
  const units: Unit[] = [];
  const troops: Troop[] = [];

  for (const army of battleState.armies) {
    let placed = 0;

    army.troops.forEach((troopSetup, troopIndex) => {
      const troopId = troopSetup.id ?? `${army.side}-t${troopIndex}`;
      const mageTypes = expand(troopSetup.mages, catalog, 'mage');
      const summonTypes = expand(troopSetup.summons, catalog, 'summon');
      assertSupported(mageTypes, summonTypes, troopId);

      const troop: Troop = { id: troopId, side: army.side, mageIds: [], summonIds: [] };

      [...mageTypes, ...summonTypes].forEach((type, memberIndex) => {
        const unitId = `${troopId}-u${memberIndex}`;
        const unit = instantiate(
          type,
          unitId,
          army.side,
          troopId,
          deploymentSpot(map, army.side, placed, rng),
        );
        placed += 1;

        units.push(unit);
        if (type.kind === 'mage') {
          troop.mageIds.push(unitId);
        } else {
          troop.summonIds.push(unitId);
        }
      });

      troops.push(troop);
    });
  }

  const duplicate = units.find((unit, index) => units.findIndex((other) => other.id === unit.id) !== index);
  if (duplicate) {
    throw new Error(`two units share the id ${duplicate.id}; troop ids must be unique`);
  }

  return {
    tick: 0,
    rngState: rng.state,
    units,
    troops,
    bases: {
      north: { maxHp: map.bases.north.maxHp, position: map.bases.north.position, hp: map.bases.north.maxHp },
      south: { maxHp: map.bases.south.maxHp, position: map.bases.south.position, hp: map.bases.south.maxHp },
    },
    zoneScore: { north: 0, south: 0 },
  };
}

File: api/app/sim/world.py
"""The world model, and how a battle's opening state is built from roster data.

The roster is a **multiset** (design doc §4.1): "3x Ember Adept, 4x Cinder Hound"
is the normal shape, and nothing in the sim assumes an army size. Round 1 fields
three mages and round 4 fields most of the roster, so any code here that
hard-coded a count would be wrong by round 2.

A troop is one or more mages plus the summons they support (§4.2). Support is
mandatory and local — a mage in another troop is too far away to help — so a
summon whose school no mage in *its own* troop supports is rejected at build time
rather than quietly standing on the field.

Placement is derived, never supplied. Every troop carries exactly one order, and
`formation.py` turns that order into a shape, a starting spot in the deployment
strip, and the station each unit walks to. Nothing in this module — or anywhere
else in the API — accepts a position, a stance or a facing from a plan.
"""

from __future__ import annotations

from collections.abc import Sequence
from dataclasses import dataclass, field

from app.sim.formation import (
    Formation,
    deployment_anchor,
    deployment_band,
    deployment_spot,
    derive_formation,
    station,
)
from app.sim.map import MapConfig, validate_map_config
from app.sim.orders import Order, validate_order
from app.sim.rng import Rng
from app.sim.schools import School
from app.sim.types import SIDES, Side, TroopId, UnitId, UnitRef, Vec2
from app.sim.units import UnitKind, UnitType, UnitTypeCatalog, build_unit_type_catalog


@dataclass
class RosterEntry:
    """One line of a roster: a card and how many copies. Omitting count means one."""

    type_id: str
    count: int = 1


@dataclass
class TroopSetup:
    """A troop as a plan gives it: a roster and the one order it is under.

    `order` comes first and has no default because it is not optional — a troop
    without an order has nothing to derive a position from, and a default would
    quietly become the placement input this design does not have.
    """

    order: Order
    mages: list[RosterEntry] = field(default_factory=list)
    summons: list[RosterEntry] = field(default_factory=list)
    id: str | None = None


@dataclass
class ArmySetup:
    side: Side
    troops: list[TroopSetup] = field(default_factory=list)


@dataclass
class BattleSetup:
    """Everything a battle needs to exist, as data."""

    unit_types: list[UnitType] = field(default_factory=list)
    armies: list[ArmySetup] = field(default_factory=list)
    #: Base HP carried in from earlier rounds. A side left out opens at full.
    #: Damage persists across a match: nothing here ever refills a base (JQ-187).
    base_hp: dict[Side, float] = field(default_factory=dict)


@dataclass
class Unit:
    id: UnitId
    type_id: str
    kind: UnitKind
    schools: tuple[School, ...]
    side: Side
    troop_id: TroopId
    max_hp: float
    damage: float
    range: float
    #: Map units per second; the loop converts to per-tick.
    speed: float
    attack_cooldown_seconds: float
    hp: float
    position: Vec2
    #: This unit's slot in its troop's formation, relative to the objective.
    #: Derived from the troop's order at battle start; never supplied.
    formation_offset: Vec2
    #: Where the unit is trying to stand this tick. The orders phase rewrites it
    #: from the troop's order every tick, so a behaviour layer (JQ-296/328) can
    #: overwrite it for a bounded diversion and get the assigned station back by
    #: simply stopping — returning to post needs no bookkeeping of its own.
    destination: Vec2
    #: Ticks still to wait before this unit can attack again — ticks, not seconds.
    cooldown_remaining: int = 0


@dataclass
class Troop:
    id: TroopId
    side: Side
    #: Exactly one, for the whole battle. The plan phase is where it is chosen.
    order: Order
    #: Living mages. A troop whose last mage dies dissolves — slice D (§4.6).
    mage_ids: list[UnitId] = field(default_factory=list)
    summon_ids: list[UnitId] = field(default_factory=list)


@dataclass
class BaseState:
    max_hp: float
    position: Vec2
    hp: float


@dataclass
class World:
    tick: int
    #: Serialisable: a snapshot fully determines every draw that follows it.
    rng_state: int
    units: list[Unit]
    troops: list[Troop]
    bases: dict[Side, BaseState]
    zone_score: dict[Side, float]
    #: Who currently holds each zone, by zone id. None while empty or contested.
    zone_holders: dict[str, Side | None]


def unit_ref(unit: Unit) -> UnitRef:
    """How a unit appears in the event stream."""
    return UnitRef(unit.id, unit.troop_id, unit.side, unit.type_id)


def is_alive(unit: Unit) -> bool:
    """A unit brought to zero stops acting at once, and is swept at end of tick."""
    return unit.hp > 0


def _expand(entries: Sequence[RosterEntry], catalog: UnitTypeCatalog, kind: UnitKind) -> list[UnitType]:
    expanded: list[UnitType] = []

    for entry in entries:
        unit_type = catalog.get(entry.type_id)
        if unit_type is None:
            raise ValueError(f"roster names {entry.type_id}, which is not in the unit type catalog")
        if unit_type.kind != kind:
            raise ValueError(f"{entry.type_id} is a {unit_type.kind}, but the roster lists it as a {kind}")
        if not isinstance(entry.count, int) or entry.count < 1:
            raise ValueError(f"roster entry {entry.type_id} has a count of {entry.count!r}")
        expanded.extend([unit_type] * entry.count)

    return expanded


def _assert_one_army_per_side(armies: Sequence[ArmySetup]) -> None:
    seen: set[Side] = set()
    for army in armies:
        if army.side in seen:
            raise ValueError(f"two armies were given for {army.side}")
        seen.add(army.side)
    for side in SIDES:
        if side not in seen:
            raise ValueError(f"no army was given for {side}")


def _assert_supported(mages: Sequence[UnitType], summons: Sequence[UnitType], troop_id: TroopId) -> None:
    if not mages:
        raise ValueError(f"troop {troop_id} has no mage, so it could hold nothing on the field")

    # Membership only — never iterated. See the note in `rng.py`.
    supported: set[School] = {school for mage in mages for school in mage.schools}
    for summon in summons:
        if not any(school in supported for school in summon.schools):
            raise ValueError(
                f"troop {troop_id} has no mage able to support {summon.id} "
                f"({'/'.join(summon.schools)}); support is mandatory and local"
            )


def _carried_base_hp(config: MapConfig, side: Side, carried: dict[Side, float]) -> float:
    """The HP this side's base opens on: what it has left, or full if unstated.

    A base never recovers between rounds (JQ-187), so a match engine hands last
    round's remaining HP straight back in.
    """
    if side not in carried:
        return config.bases[side].max_hp

    hp = carried[side]
    if not isinstance(hp, (int, float)) or isinstance(hp, bool):
        raise ValueError(f"carried {side} base HP must be a number, got {hp!r}")
    if hp <= 0:
        raise ValueError(f"{side} base is carried in at {hp} HP, so the match is already over")
    if hp > config.bases[side].max_hp:
        raise ValueError(f"{side} base is carried in above its maximum; bases do not recover")
    return float(hp)


def create_world(config: MapConfig, battle_state: BattleSetup, rng: Rng) -> World:
    """Builds the opening state of a battle. Pure: same inputs, same world."""
    validate_map_config(config)
    _assert_one_army_per_side(battle_state.armies)

    catalog = build_unit_type_catalog(battle_state.unit_types)
    units: list[Unit] = []
    troops: list[Troop] = []
    seen_ids: set[UnitId] = set()

    for army in battle_state.armies:
        for troop_index, troop_setup in enumerate(army.troops):
            troop_id = troop_setup.id or f"{army.side}-t{troop_index}"
            order = troop_setup.order
            validate_order(order, config)

            mage_types = _expand(troop_setup.mages, catalog, "mage")
            summon_types = _expand(troop_setup.summons, catalog, "summon")
            _assert_supported(mage_types, summon_types, troop_id)

            band = deployment_band(config, army.side, troop_index, len(army.troops))
            formation: Formation = derive_formation(
                order,
                army.side,
                len(mage_types),
                len(summon_types),
                band.end - band.start,
            )
            anchor = deployment_anchor(config, army.side, order, band, formation)
            troop = Troop(id=troop_id, side=army.side, order=order)

            for member_index, unit_type in enumerate([*mage_types, *summon_types]):
                unit_id = f"{troop_id}-u{member_index}"
                if unit_id in seen_ids:
                    raise ValueError(f"two units share the id {unit_id}; troop ids must be unique")
                seen_ids.add(unit_id)

                offset = formation.offsets[member_index]
                units.append(
                    Unit(
                        id=unit_id,
                        type_id=unit_type.id,
                        kind=unit_type.kind,
                        schools=unit_type.schools,
                        side=army.side,
                        troop_id=troop_id,
                        max_hp=unit_type.max_hp,
                        damage=unit_type.damage,
                        range=unit_type.range,
                        speed=unit_type.speed,
                        attack_cooldown_seconds=unit_type.attack_cooldown_seconds,
                        hp=unit_type.max_hp,
                        position=deployment_spot(config, army.side, anchor, offset, rng),
                        formation_offset=offset,
                        destination=station(order, army.side, offset, config),
                    )
                )

                if unit_type.kind == "mage":
                    troop.mage_ids.append(unit_id)
                else:
                    troop.summon_ids.append(unit_id)

            troops.append(troop)

    return World(
        tick=0,
        rng_state=rng.state,
        units=units,
        troops=troops,
        bases={
            side: BaseState(
                max_hp=config.bases[side].max_hp,
                position=config.bases[side].position,
                hp=_carried_base_hp(config, side, battle_state.base_hp),
            )
            for side in SIDES
        },
        zone_score={"north": 0, "south": 0},
        zone_holders={zone.id: None for zone in config.zones},
    )


def troop_of(world: World, unit: Unit) -> Troop:
    """The troop a unit belongs to. Every unit belongs to exactly one."""
    for troop in world.troops:
        if troop.id == unit.troop_id:
            return troop
    raise ValueError(f"unit {unit.id} names troop {unit.troop_id}, which is not in this battle")


def order_of(world: World, unit: Unit) -> Order:
    """The order this unit is acting under — its troop's, always."""
    return troop_of(world, unit).order


def orders_by_troop(world: World) -> dict[TroopId, Order]:
    """Every troop's order, for a phase that needs to look one up per unit.

    A dict keyed by troop id: looked up, never iterated. Iterating it would order
    the sim by Python's per-process string hashing. See `rng.py`.
    """
    return {troop.id: troop.order for troop in world.troops}
